internal/ingress/types.go

```diff
@@ -85,6 +85,30 @@ type Backend struct {
 	UpstreamHashBy string `json:"upstream-hash-by,omitempty"`
 	// LB algorithm configuration per ingress
 	LoadBalancing string `json:"load-balance,omitempty"`
+	// Denotes if a backend is virtual. A virtual backend is not backed by it's own server and instead shares
+	// a server with another, non-virtual backend. This can be used to share multiple upstreams in the same
+	// nginx server block.
+	Virtual bool `json:"virtual"`
+	// Additional metadata to describe the characteristics of a virtual backend.
+	// +optional
+	VirtualMetadata VirtualMetadata `json:"virtualMetadata,omitempty"`
```

Comment:
@ElvinEfendi How is this rather than the `VirtualBackend` struct? I think this is better since it also means the whole backend is initialized in the `createUpstreams` function rather than initializing virtual backends in the `mergeVirtualBackends` function.

I agree in thinking using the name `VirtualBackend` isn't very appropriate.  A true `Virtual` backend would be better represented as the top level backend being virtual and it contains a list of real backends. Perhaps something like:

```
type Backend struct {
  // Whether or not the backend has a server
  HasServer bool
  TrafficShapingConfig TrafficShapingConfig
  // List of backends that requests could also be routed to
  AlternativeBackends []*Backend
}

// Contains rules for how to route traffic to alternative backends
type TrafficShapingConfig struct {
  Weight int  
  Header string
  Cookie string
}
```

This drops the virtual connotations while being clear that a backend can standalone or not. I think a better name could be come created than `TrafficShapingConfig`, though 🤔

Reply:
I too have difficulties with the `Virtual` naming.

What you're trying to do is splitting routing into multiple backends, which allows us then to do some Canary. But we should not limit it to the Canary feature.

Naming is hard, but I prefer this option over the `Virtual` one.